File: x402-crewai-agent/utils/tools_registry.py
"""
Tool Registry - Easy way to add paid and free APIs for CrewAI agents

To add a new paid API:
1. Add entry to PAID_APIS list
2. That's it! Agent will automatically use x402 for payment
"""

from typing import Dict, Any, List, Optional, Callable
from pydantic import BaseModel, Field
from enum import Enum


class HTTPMethod(str, Enum):
    GET = "GET"
    POST = "POST"
    PUT = "PUT"
    DELETE = "DELETE"


class APIParameter(BaseModel):
    """Parameter definition for API calls"""
    name: str
    type: str  # "string", "number", "boolean", "array", "object"
    description: str
    required: bool = True
    default: Any = None


class APIConfig(BaseModel):
    """Configuration for a paid or free API"""
    name: str
    description: str
    endpoint: str
    method: HTTPMethod = HTTPMethod.GET
    cost: float = 0.0
    parameters: List[APIParameter] = []
    transform: Optional[Callable] = None  # Transform parameters before sending
    build_url: Optional[Callable] = None  # Build URL with query params (for GET)
    headers: Dict[str, str] = {}

    class Config:
        arbitrary_types_allowed = True


"""
Define your paid APIs here
Add as many as you want - they'll be available as tools to the agent
"""
PAID_APIS = [
    APIConfig(
        name="test_sentiment_analysis",
        description="Test sentiment analysis with transaction tracking (costs $0.10 USD) - Working demo!",
        endpoint="http://localhost:3000/api/test-paid",
        method=HTTPMethod.POST,
        cost=0.10,
        parameters=[
            APIParameter(
                name="text",
                type="string",
                description="The text to analyze",
                required=True
            )
        ],
        transform=lambda params: {"text": params.get("text")}
    ),

    APIConfig(
        name="weather",
        description="Get current weather for a city (costs $0.10 USD)",
        endpoint="http://localhost:8000/api/weather",
        method=HTTPMethod.GET,
        cost=0.10,
        parameters=[
            APIParameter(
                name="city",
                type="string",
                description="City name to get weather for",
                required=True
            )
        ],
        build_url=lambda params: f"http://localhost:8000/api/weather?city={params.get('city')}"
    ),

    APIConfig(
        name="stock_data",
        description="Get stock market data for a symbol (costs $0.25 USD)",
        endpoint="http://localhost:8000/api/stock_data",
        method=HTTPMethod.GET,
        cost=0.25,
        parameters=[
            APIParameter(
                name="symbol",
                type="string",
                description="Stock symbol (e.g., AAPL, GOOGL)",
                required=True
            )
        ],
        build_url=lambda params: f"http://localhost:8000/api/stock_data?symbol={params.get('symbol')}"
    ),

    APIConfig(
        name="news",
        description="Get latest news articles on any topic (costs $0.15 USD)",
        endpoint="http://localhost:8000/api/news",
        method=HTTPMethod.GET,
        cost=0.15,
        parameters=[
            APIParameter(
                name="topic",
                type="string",
                description="News topic to search for",
                required=True
            )
        ],
        build_url=lambda params: f"http://localhost:8000/api/news?topic={params.get('topic')}"
    ),

    APIConfig(
        name="translation",
        description="Translate text to target language (costs $0.20 USD)",
        endpoint="http://localhost:8000/api/translation",
        method=HTTPMethod.POST,
        cost=0.20,
        parameters=[
            APIParameter(
                name="text",
                type="string",
                description="Text to translate",
                required=True
            ),
            APIParameter(
                name="target_language",
                type="string",
                description="Target language code (e.g., 'es', 'fr', 'de')",
                required=True
            )
        ],
        transform=lambda params: {
            "text": params.get("text"),
            "target_language": params.get("target_language")
        }
    ),

    APIConfig(
        name="data_analysis",
        description="Perform comprehensive data analysis (costs $0.50 USD)",
        endpoint="http://localhost:8000/api/data_analysis",
        method=HTTPMethod.POST,
        cost=0.50,
        parameters=[
            APIParameter(
                name="data",
                type="object",
                description="Data object to analyze",
                required=True
            )
        ],
        transform=lambda params: params.get("data", {})
    ),

    APIConfig(
        name="paid_weather_x402",
        description="Get weather with full x402 payment flow (costs $0.10 USD) - Uses x402 protocol",
        endpoint="https://2701e145a0b0.ngrok-free.app/api/weather",
        method=HTTPMethod.GET,
        cost=0.10,
        parameters=[
            APIParameter(
                name="latitude",
                type="number",
                description="Latitude coordinate",
                required=True
            ),
            APIParameter(
                name="longitude",
                type="number",
                description="Longitude coordinate",
                required=True
            )
        ],
        build_url=lambda params: f"https://2701e145a0b0.ngrok-free.app/api/weather?latitude={params.get('latitude')}&longitude={params.get('longitude')}"
    ),
]


"""
Define your free APIs here
"""
FREE_APIS = [
    APIConfig(
        name="search_web",
        description="Search the web using DuckDuckGo (free)",
        endpoint="https://api.duckduckgo.com/",
        method=HTTPMethod.GET,
        cost=0.0,
        parameters=[
            APIParameter(
                name="query",
                type="string",
                description="Search query",
                required=True
            )
        ],
        build_url=lambda params: f"https://api.duckduckgo.com/?q={params.get('query')}&format=json"
    ),

    APIConfig(
        name="free_joke",
        description="Get a random joke (free)",
        endpoint="https://official-joke-api.appspot.com/random_joke",
        method=HTTPMethod.GET,
        cost=0.0,
        parameters=[]
    ),
]


def get_all_apis() -> List[APIConfig]:
    """Get all available APIs (paid + free)"""
    return PAID_APIS + FREE_APIS


def get_api_by_name(name: str) -> Optional[APIConfig]:
    """Get API configuration by name"""
    for api in get_all_apis():
        if api.name == name:
            return api
    return None


def get_paid_apis() -> List[APIConfig]:
    """Get only paid APIs"""
    return PAID_APIS


def get_free_apis() -> List[APIConfig]:
    """Get only free APIs"""
    return FREE_APIS


def calculate_cost(api_names: List[str]) -> float:
    """Calculate total cost for multiple API calls"""
    total = 0.0
    for name in api_names:
        api = get_api_by_name(name)
        if api:
            total += api.cost
    return total


def get_api_summary() -> Dict[str, Any]:
    """Get summary of all available APIs"""
    paid_apis = get_paid_apis()
    free_apis = get_free_apis()

    return {
        "total_apis": len(get_all_apis()),
        "paid_apis": len(paid_apis),
        "free_apis": len(free_apis),
        "total_paid_cost": sum(api.cost for api in paid_apis),
        "api_list": {
            "paid": [{"name": api.name, "cost": api.cost, "description": api.description} for api in paid_apis],
            "free": [{"name": api.name, "description": api.description} for api in free_apis]
        }
    }

File: vercel-ai-x402-agent/lib/tools-registry.ts
/**
 * Tool Registry - Easy way to add paid and free APIs
 *
 * To add a new paid API:
 * 1. Add entry to PAID_APIS array
 * 2. That's it! Agent will automatically use x402 for payment
 */

import { z } from 'zod';
import { X402ServerClient } from './x402-server';
import { getBaseApiUrl } from './config';

// Initialize x402 client for agent payments
const x402Client = new X402ServerClient(
  process.env.AGENT_WALLET_PRIVATE_KEY || ''
);

/**
 * Define your paid APIs here
 * Add as many as you want - they'll be available as tools to the agent
 */
export const PAID_APIS = [
  {
    name: 'test_sentiment_analysis',
    description: 'Test sentiment analysis with transaction tracking (costs $0.10 USDC) - Working demo!',
    endpoint: `${getBaseApiUrl()}/api/test-paid`,
    method: 'POST' as const,
    cost: 0.10,
    schema: z.object({
      text: z.string().describe('The text to analyze')
    }),
    transform: (params: any) => ({
      text: params.text
    })
  },
  {
    name: 'premium_sentiment_analysis',
    description: 'Advanced AI-powered sentiment analysis using GPT (costs $0.10 USDC) - Real OpenAI API!',
    endpoint: `${getBaseApiUrl()}/api/sentiment`,
    method: 'POST' as const,
    cost: 0.10,
    schema: z.object({
      text: z.string().describe('The text to analyze'),
      language: z.string().optional().describe('Language code (default: en)')
    }),
    transform: (params: any) => ({
      text: params.text,
      language: params.language || 'en'
    })
  },
  {
    name: 'data_enrichment',
    description: 'Enrich company data with premium database (costs $0.25 USDC)',
    endpoint: 'https://api.example.com/enrich',
    method: 'POST' as const,
    cost: 0.25,
    schema: z.object({
      company_name: z.string().describe('Company name to enrich'),
      fields: z.array(z.string()).describe('Fields to retrieve: revenue, employees, industry, etc.')
    }),
    transform: (params: any) => ({
      company: params.company_name,
      fields: params.fields
    })
  },
  {
    name: 'advanced_translation',
    description: 'Professional translation with context awareness (costs $0.15 USDC per 1000 chars)',
    endpoint: 'https://api.example.com/translate',
    method: 'POST' as const,
    cost: 0.15,
    schema: z.object({
      text: z.string().describe('Text to translate'),
      source_lang: z.string().describe('Source language code'),
      target_lang: z.string().describe('Target language code'),
      context: z.string().optional().describe('Additional context for better translation')
    }),
    transform: (params: any) => params
  },
  {
    name: 'advanced_translation',
    description: 'Professional translation with context awareness (costs $0.15 USDC per 1000 chars)',
    endpoint: 'https://api.example.com/translate',
    method: 'POST' as const,
    cost: 0.15,
    schema: z.object({
      text: z.string().describe('Text to translate'),
      source_lang: z.string().describe('Source language code'),
      target_lang: z.string().describe('Target language code'),
      context: z.string().optional().describe('Additional context for better translation')
    }),
    transform: (params: any) => params
  },
  {
    name: 'paid_weather',
    description: 'Get current weather for a location with 402 payment flow (costs $0.10 USDC)',
    endpoint: 'https://2701e145a0b0.ngrok-free.app/api/weather',
    method: 'GET' as const,
    cost: 0.1,
    schema: z.object({
      latitude: z.number().describe('Latitude'),
      longitude: z.number().describe('Longitude')
    }),
    transform: (params: any) => null, // GET request, params in URL
    buildUrl: (params: any) =>
      `https://2701e145a0b0.ngrok-free.app/api/weather?latitude=${params.latitude}&longitude=${params.longitude}`
  },
];

/**
 * Define your free APIs here
 */
export const FREE_APIS = [
  {
    name: 'search_web',
    description: 'Search the web using DuckDuckGo (free)',
    endpoint: 'https://api.duckduckgo.com/',
    method: 'GET' as const,
    cost: 0,
    schema: z.object({
      query: z.string().describe('Search query')
    }),
    transform: (params: any) => null,
    buildUrl: (params: any) =>
      `https://api.duckduckgo.com/?q=${encodeURIComponent(params.query)}&format=json`
  }
];

/**
 * Execute a paid API call using x402
 */
export async function executePaidAPI(
  apiConfig: typeof PAID_APIS[0],
  params: any,
  userId?: string
): Promise<any> {
  console.log(`🔐 Executing paid API: ${apiConfig.name} (cost: $${apiConfig.cost})`);

  // TODO: In production, check user balance here
  // const user = await db.users.findOne({ id: userId });
  // if (user.balance < apiConfig.cost) {
  //   throw new Error(`Insufficient balance. Required: $${apiConfig.cost}, Available: $${user.balance}`);
  // }

  try {
    // Build URL if buildUrl function exists (for GET requests)
    const url = (apiConfig as any).buildUrl ? (apiConfig as any).buildUrl(params) : apiConfig.endpoint;

    // Transform parameters
    const body = apiConfig.transform(params);

    // Make paid API call using x402
    const response = await x402Client.fetch(url, {
      method: apiConfig.method,
      headers: {
        'Content-Type': 'application/json'
      },
      body: apiConfig.method === 'GET' ? undefined : JSON.stringify(body)
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`API call failed: ${error}`);
    }

    const result = await response.json();

    // Log the API response
    console.log('📊 API Response:', JSON.stringify(result, null, 2));

    // Get transaction info
    const txInfo = x402Client.getLastTransaction();

    // TODO: In production, deduct from user balance
    // await db.users.updateOne(
    //   { id: userId },
    //   { $inc: { balance: -apiConfig.cost } }
    // );

    console.log(`✅ Paid API call successful`);

    return {
      ...result,
      _transaction: txInfo.hash ? {
        hash: txInfo.hash,
        network: txInfo.network,
        explorerUrl: getExplorerUrl(txInfo.hash, txInfo.network)
      } : null
    };

  } catch (error) {
    console.error(`❌ Paid API call failed:`, error);
    throw error;
  }
}

/**
 * Get block explorer URL for transaction
 */
function getExplorerUrl(txHash: string, network: string | null): string {
  const explorers: Record<string, string> = {
    'base': 'https://basescan.org/tx/',
    'base-sepolia': 'https://sepolia.basescan.org/tx/',
    'ethereum': 'https://etherscan.io/tx/',
    'polygon': 'https://polygonscan.com/tx/',
    'arbitrum': 'https://arbiscan.io/tx/',
    'optimism': 'https://optimistic.etherscan.io/tx/'
  };

  const baseUrl = explorers[network || 'base-sepolia'] || explorers['base-sepolia'];
  return `${baseUrl}${txHash}`;
}

/**
 * Execute a free API call
 */
export async function executeFreeAPI(
  apiConfig: typeof FREE_APIS[0],
  params: any
): Promise<any> {
  console.log(`🆓 Executing free API: ${apiConfig.name}`);

  try {
    const url = apiConfig.buildUrl ? apiConfig.buildUrl(params) : apiConfig.endpoint;

    const response = await fetch(url, {
      method: apiConfig.method,
      headers: {
        'Content-Type': 'application/json'
      }
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`API call failed: ${error}`);
    }

    const result = await response.json();
    console.log(`✅ Free API call successful`);

    return result;

  } catch (error) {
    console.error(`❌ Free API call failed:`, error);
    throw error;
  }
}

/**
 * Get all available tools (paid + free) for the agent
 */
export function getAllAPIs() {
  return [...PAID_APIS, ...FREE_APIS];
}

/**
 * Get tool by name
 */
export function getAPIByName(name: string) {
  return getAllAPIs().find(api => api.name === name);
}

/**
 * Calculate total cost for multiple API calls
 */
export function calculateCost(apiNames: string[]): number {
  return apiNames.reduce((total, name) => {
    const api = getAPIByName(name);
    return total + (api?.cost || 0);
  }, 0);
}
